grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -86,57 +89,76 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, GraknTx graph
 
-        Collection<Set<Fragment>> connectedFragmentSets = getConnectedFragmentSets(query, allNodes);
+        final Set<Node> connectedNodes = new HashSet<>();
+        final Map<Node, Double> nodesWithFixedCost = new HashMap<>();
 
-        connectedFragmentSets.forEach(fragmentSet -> {
+//        Collection<Set<Fragment>> connectedFragmentSets =
+        getConnectedFragmentSets(plan, query, allNodes, connectedNodes, nodesWithFixedCost, tx).forEach(fragmentSet -> {
 
-            Set<Node> connectedNodes = new HashSet<>();
             Map<Node, Map<Node, Fragment>> edges = new HashMap<>();
-            final Set<Node> nodesWithFixedCost = new HashSet<>();
             Set<Weighted<DirectedEdge<Node>>> weightedGraph = new HashSet<>();
 
-            fragmentSet.stream()
-                    .filter(filterNodeFragment(plan, allNodes, connectedNodes, nodesWithFixedCost))
+            final Set<Node> startingNodeSet = new HashSet<>();
+            final Set<Fragment> edgeFragmentSet = new HashSet<>();
+
+            fragmentSet.forEach(fragment -> {
+                if (fragment.end() != null) {
+                    edgeFragmentSet.add(fragment);
+                } else if (fragment.hasFixedFragmentCost()) {
+                    startingNodeSet.add(Node.addIfAbsent(NodeId.NodeType.VAR, fragment.start(), allNodes));
+                }
+            });
+
+            Set<Weighted<DirectedEdge<Node>>> edgeSet = edgeFragmentSet.stream()
                     .flatMap(fragment -> fragment.directedEdges(allNodes, edges).stream())
-                    .forEach(weightedDirectedEdge -> {
-                        weightedGraph.add(weightedDirectedEdge);
-                        connectedNodes.add(weightedDirectedEdge.val.destination);
-                        connectedNodes.add(weightedDirectedEdge.val.source);
-                    });
+                    .collect(Collectors.toSet()); // make sure previous steps are all done
+
+            edgeSet.forEach(weightedDirectedEdge -> {
+                if (nodesWithFixedCost.containsKey(weightedDirectedEdge.val.source) &&
+                        nodesWithFixedCost.get(weightedDirectedEdge.val.source) > 0 &&
+                        weightedDirectedEdge.val.destination.getNodeId().getNodeType() == NodeId.NodeType.ISA) {
+
+                    edges.get(weightedDirectedEdge.val.destination).get(weightedDirectedEdge.val.source)
+                            .setAccurateFragmentCost(nodesWithFixedCost.get(weightedDirectedEdge.val.source));
+                    weightedDirectedEdge.weight = -nodesWithFixedCost.get(weightedDirectedEdge.val.source);
+                    startingNodeSet.add(weightedDirectedEdge.val.source);
+                }
+                weightedGraph.add(weightedDirectedEdge);
+                connectedNodes.add(weightedDirectedEdge.val.destination);
+                connectedNodes.add(weightedDirectedEdge.val.source);
+            });
```

Comment:
It might be a good idea to split this method up plus add some clarifying variables and comments.